### ISSUE TITLE
🐛 Bug Report: Routing error on default catalog page (at least for Systems)

### ISSUE THREAD

**Issue by efrenaguilar95** [backstage/backstage#27894]
### 📜 Description

Just update my Backstage instance from v1.31 to v1.33. So I believe this bug comes from this change; however, not 100% sure.

When viewing the catalog page for a "System"entity, there is a tab on the type for the "Diagram" of the entity. Clicking on this messes with the routing for future navigation, adding "diagram" to the end of all of them.

For example, once I navigate to the diagram tab, navigating back to the "Overview" gives the route "system/{system-name}/diagram". The diagram also has a bugged route of "system/{system-name}/diagram/diagram"

### 👀 Have you spent some time to check if this bug has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/backstage/backstage/blob/master/CODE_OF_CONDUCT.md)

### Are you willing to submit PR?

No, I don't have time to work on this right now

**Comment by jizusun** [comment]
Hi @efrenaguilar95  I would like to help, and can you help attach a screenshot or two for this bug?

**Comment by Rugvip** [comment]
@efrenaguilar95 please provide reproduction steps or move information. There's no open source plugin that I'm aware of with a `diagram` route so this may be an issue in a custom plugin or app wiring. Still interested to figure out what leads to this bug though

**Comment by efrenaguilar95** [comment]
Sure, here are the repo steps. Hope this helps, apologies, I am running this on a company computer and just found a solution to correctly getting screenshots from there without getting tapped on the shoulder.

1. On the entity page for a System entity. Clicking the "Diagram" page

![image](https://github.com/user-attachments/assets/c6b4212b-9a17-4adc-b246-f408ddadfd94)

2. Diagram page loads correctly.

![image](https://github.com/user-attachments/assets/c96ed044-6f8e-4cee-ac04-13fab87e5e7a)

3. Attempting to navigate back to "Overview" incorrectly routes. See address on bottom

![image](https://github.com/user-attachments/assets/1bcd9acb-3f64-44cb-b5b6-570b29324af7)

4. Similarly, the diagram tab is now also bugged with routing.

![image](https://github.com/user-attachments/assets/ce751cbe-2ed7-4d43-8d91-737dd21c23d2)

**Comment by efrenaguilar95** [comment]
Also, this is using the same format of System page as that generated from a new Backstage project.

Specifically, here is the definition. Hope this helps, perhaps one of these components is the source of the error

```js
const systemPage = (
  <EntityLayout>
    <EntityLayout.Route path="/" title="Overview">
      <Grid container spacing={3} alignItems="stretch">
        {entityWarningContent}
        <Grid item md={6}>
          <EntityAboutCard variant="gridItem" />
        </Grid>
        <Grid item md={6} xs={12}>
          <EntityCatalogGraphCard variant="gridItem" height={400} />
        </Grid>
        <Grid item md={4} xs={12}>
          <EntityLinksCard />
        </Grid>
        <Grid item md={8}>
          <EntityHasComponentsCard variant="gridItem" />
        </Grid>
        <Grid item md={6}>
          <EntityHasApisCard variant="gridItem" />
        </Grid>
        <Grid item md={6}>
          <EntityHasResourcesCard variant="gridItem" />
        </Grid>
      </Grid>
    </EntityLayout.Route>
    <EntityLayout.Route path="/diagram" title="Diagram">
      <EntityCatalogGraphCard
        variant="gridItem"
        direction={Direction.TOP_BOTTOM}
        title="System Diagram"
        height={700}
        relations={[
          RELATION_PART_OF,
          RELATION_HAS_PART,
          RELATION_API_CONSUMED_BY,
          RELATION_API_PROVIDED_BY,
          RELATION_CONSUMES_API,
          RELATION_PROVIDES_API,
          RELATION_DEPENDENCY_OF,
          RELATION_DEPENDS_ON,
        ]}
        unidirectional={false}
      />
    </EntityLayout.Route>
  </EntityLayout>
);
```

**Comment by awanlin** [comment]
Just adding on that we don't have this issue on the Demo site. The list of Systems is here: https://demo.backstage.io/catalog?filters%5Bkind%5D=system&filters%5Bnamespace%5D=default&filters%5Buser%5D=all&limit=20. Opening any of these and then clicking the Diagram tabs loads with the expected URL and contents. Navigating back to the Overview tabs also loads the content with the expected URL.

**Comment by efrenaguilar95** [comment]
> Just adding on that we don't have this issue on the Demo site. The list of Systems is here: https://demo.backstage.io/catalog?filters%5Bkind%5D=system&filters%5Bnamespace%5D=default&filters%5Buser%5D=all&limit=20. Opening any of these and then clicking the Diagram tabs loads with the expected URL and contents. Navigating back to the Overview tabs also loads the content with the expected URL.

Hmmm, interesting. Good to know.

I also just realized the error is more widespread than I thought. This occurs on all tabs for me (don't know how I missed that).

**Comment by awanlin** [comment]
Ok, so this new information about the error is helpful. I would say almost a year ago we had several issues logged about the routes being duplicated, this ended up being caused by a bad release or unintended change in `react-router`. I'm wondering if maybe that's what you are hitting.

**Comment by awanlin** [comment]
Found the issue faster than expected, is this what you are experiencing? https://github.com/backstage/backstage/issues/21487

**Comment by efrenaguilar95** [comment]
> Found the issue faster than expected, is this what you are experiencing? #21487

Yup that's exactly it, except my browser does not add on the route/won't navigate to it. I'm assuming if it did, the paths would keep appending as shown in that issue.

**Comment by efrenaguilar95** [comment]
Huh, almost the same exact time as last year too.

**Comment by awanlin** [comment]
Can you share what version of `react-router` you are using? Run `yarn why react-router` and post the results, please.

**Comment by efrenaguilar95** [comment]
$ yarn why react-router
├─ app@workspace:packages/app
│  └─ react-router@npm:7.0.1 [f87a9] (via npm:^7.0.1 [f87a9])
│
├─ react-router-dom@npm:7.0.1
│  └─ react-router@npm:7.0.1 (via npm:7.0.1)
│
└─ react-router-dom@npm:7.0.1 [f87a9]
   └─ react-router@npm:7.0.1 [f87a9] (via npm:^7.0.1 [f87a9])

**Comment by awanlin** [comment]
Yeah, that would do it. I don't think we are in a place to support v7 yet. Pretty sure that the fix for the issue from a year ago moved the logic behind a feature flag and that flag is now the default in v7 which is why you are seeing this issue again.

I would try using a resolution in your root `package.json` to pin `react-router` to `6.28.0`

**Comment by efrenaguilar95** [comment]
> Yeah, that would do it. I don't think we are in a place to support v7 yet. Pretty sure that the fix for the issue from a year ago moved the logic behind a feature flag and that flag is now the default in v7 which is why you are seeing this issue again.
> 
> I would try using a resolution in your root `package.json` to pin `react-router` to `6.28.0`

Yup, pinning that along with `react-router-dom` did the trick. Not sure if we should leave this open, or replace with a different issue.

**Comment by awanlin** [comment]
I've logged an issue for the migration to v7 of React Router to help track this: https://github.com/backstage/backstage/issues/27927

**Comment by mattthym** [comment]
I can confirm that pinning `react-router` and `react-router-dom` to `6.28.0` fixes the routing issue at hand :heavy_check_mark:

**Comment by awanlin** [comment]
Hi @mattthym, any chance you can share how you managed to get onto v7? Did you upgrade intentionally? Just trying to understand how people are ending up here.

**Comment by mattthym** [comment]
@awanlin Sure, can do!
So I first upgraded Backstage to the newest version with `yarn backstage-cli versions:bump`.

Since I also wanted to update the plugins, I then followed the [advice in the docs](https://backstage.io/docs/getting-started/keeping-backstage-updated/#updating-backstage-versions-with-backstage-cli) and upgraded using `--pattern`.
The docs mention to explicitely list all plugins to be updated:
```
yarn backstage-cli versions:bump --pattern '@{backstage,roadiehq}/*'
```
Since I have a couple of plugins, I felt a little lazy and upgraded with `yarn backstage-cli versions:bump --pattern '*'`.
I understand that this is why `react-router` was upgraded and that this caused the issue, so thanks again for pointing out a solution.
The resulting diff in `packages/app/package.json` also contains the upgrade of `react-router` and `react-router-dom`:
```
│ 61 │    "react-router": "^6.27.0",         │ 61 │    "react-router": "^7.0.1",
│ 62 │    "react-router-dom": "^6.27.0",     │ 62 │    "react-router-dom": "^7.0.1",
```
